v2/devdoc/logging.md

```diff
@@ -0,0 +1,339 @@
+
+The exposed API shall allow creating contexts both dynamically and on the stack and using those contexts for logging.
+
+The API shall allow easy formatting of values to be placed in the contexts by using printf-like formatting.
+
+```
+void logger_log(LOG_LEVEL log_level, LOG_CONTEXT_HANDLE log_context, const char* file, const char* func, int line_no, const char* format, ...);
+
+#define LOGGER_LOG(log_level, log_context, format, ...) \
+    logger_log(log_level, log_context, __FILE__, __FUNCTION__, __LINE__, format MU_IFCOMMALOGIC(MU_COUNT_ARG(__VA_ARGS__)) __VA_ARGS__)
+
+#define LOGGER_LOG_EX(log_level, ...) \
+    ...
+
+// macro that can be used to create a dynamically allocated context
+#define LOG_CONTEXT_CREATE(dest_log_context, parent_context, ...) \
+    ...
+
+// destroy a dynamically allocated context
+#define LOG_CONTEXT_DESTROY(log_context_handle) \
+    ...
+
+// macro that can be used to create a context on the stack.
+// stack allocated contexts do not require any destructor
+#define LOG_CONTEXT_LOCAL_DEFINE(log_context, parent_context, ...) \
+    ...
+
+// when a user wants to specify the log context type name, the user can use this macro
+// otherwise, the default is to use the context variable name as type name
+// LOG_CONTEXT_TYPE_NAME(log_context_type_name)
+
+// when a user wants to specify properties for a context, they can use LOG_CONTEXT_PROPERTY
+// LOG_CONTEXT_PROPERTY(property_type, property_name, property_value)
+
+// when a user wants to specify string fields in a context using printf-style formatting, they can use LOG_CONTEXT_STRING_PROPERTY
+// LOG_CONTEXT_STRING_PROPERTY(property_name, property_conversion_specifier, property_value)
+```
+
+Various example usages:
+
+```c
+void log_from_a_function(LOG_CONTEXT_HANDLE log_context)
+{
+    LOGGER_LOG(LOG_LEVEL_ERROR, log_context, "log from a function!");
+}
+
+void examples(void)
+{
+    char* prop_value = malloc(2);
+    (void)strcpy(prop_value, "a");
+
+    // allocate a context dynamically
+    LOG_CONTEXT_HANDLE log_context;
+    LOG_CONTEXT_CREATE(log_context, NULL, LOG_CONTEXT_STRING_PROPERTY(property_name, "%s", MU_P_OR_NULL(prop_value)));
+
+    LOGGER_LOG(LOG_LEVEL_ERROR, log_context, "some_error with context");
+
+    LOG_CONTEXT_DESTROY(log_context);
+
+    // free the string, showing that a copy is made in the context
+    free(prop_value);
+
+    // chained allocated context
+    LOG_CONTEXT_HANDLE context_1;
+    LOG_CONTEXT_CREATE(context_1, NULL, LOG_CONTEXT_NAME(context_1), LOG_CONTEXT_STRING_PROPERTY(name, "%s", "haga"));
+    LOG_CONTEXT_HANDLE context_2;
+    LOG_CONTEXT_CREATE(context_2, context_1, LOG_CONTEXT_NAME(context_2), LOG_CONTEXT_STRING_PROPERTY(last name, "%s%s", "ua", "ga"), LOG_CONTEXT_PROPERTY(int32_t, age, 42));
+    
+    LOGGER_LOG(LOG_LEVEL_CRITICAL, context_1, "log with context 1, value of some_var = %d", 42);
```

Comment:
>log with context 1 [](http://example.com/codeflow?start=47&length=18)

404, this "log with context 1" string is nowhere to be found in further output.

IS the output and the code out of sync?

Reply:
It was, has been synched and also removed the log with context 1 and log with context 2